Sources/Shared/AuthService.swift

```diff
@@ -1,78 +1,96 @@
 import Foundation
 
 @objc public class AuthService: NSObject {
-
+  
   public typealias Completion = (String?, NSError?) -> Void
-
+  
   public let name: String
   public let config: AuthConfig
   public var locker: Lockable
-
+  
   private var pendingTokenCompletions = [Completion]()
   private var executing = false
-
+  private let tokenQueue = dispatch_queue_create("OhMyAuth.AuthService.TokenQueue", DISPATCH_QUEUE_CONCURRENT)
+  
   public var tokenIsExpired: Bool {
     return locker.expiryDate?.timeIntervalSinceReferenceDate
       <= NSDate.timeIntervalSinceReferenceDate() - config.minimumValidity
   }
-
+  
   // MARK: - Initialization
-
+  
   public init(name: String, config: AuthConfig) {
     self.name = name
     self.config = config
     self.config.name = name
     locker = Locker(name: name)
   }
-
+  
   // MARK: - Authorization
-
+  
   public func authorize() -> Bool {
     guard let URL = config.authorizeURL else { return false }
-
+    
     locker.clear()
     config.webView.open(URL)
-
+    
     return true
   }
-
+  
   public func changeUser() -> Bool {
     guard let URL = config.changeUserURL else { return false }
-
+    
     locker.clear()
     config.webView.open(URL)
-
+    
     return true
   }
-
+  
   // MARK: - Tokens
-
+  
   public func accessToken(completion: Completion) {
-    guard locker.expiryDate != nil else {
-      completion(nil, Error.NoExpiryDateFound.toNSError())
-      return
-    }
-
-    guard tokenIsExpired else {
-      completion(locker.accessToken, nil)
-      return
-    }
-
-    pendingTokenCompletions.append(completion)
-
-    guard !executing else { return }
-
-    refreshToken() { [weak self] accessToken, error in
-      guard let weakSelf = self else { return }
-
-      weakSelf.pendingTokenCompletions.forEach { completion in
-        completion(accessToken, error)
+    dispatch_barrier_async(tokenQueue) { [weak self] in
```

Comment:
What is this voodoo black magic?

![0d41808f60af8871fa122b3b0f37ab1b_360 gif png](https://cloud.githubusercontent.com/assets/57446/13421331/02d8a0e8-df8e-11e5-9f8d-678819bba761.gif)